lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <nrf_inbuilt_key.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define NRF_INBUILT_KEY_OP_LS "AT%CMNG=1"
+#define NRF_INBUILT_KEY_OP_RD "AT%CMNG=2"
+#define NRF_INBUILT_KEY_OP_WR "AT%CMNG=0"
+#define NRF_INBUILT_KEY_OP_RM "AT%CMNG=3"
+
+#define AT_CMNG_PARAMS_COUNT 5
+#define AT_CMNG_CONTENT_INDEX 4
+
+static char scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	int result;
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	result = at_cmd_write("AT+CMEE?", response, sizeof(response), NULL);
+	if ((result == 0) && strlen(response) > 0) {
+		if (strchr(&response[6], 1)) {
+			return 1;
+		} else {
+			return 0;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	int result;
+
+	result = at_cmd_write(at_cmee_strings[(int)enable], NULL, 0, NULL);
+
+	if (result != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+static int write_at_cmd_with_cme_enabled(char *cmd, char *buf, size_t buf_len,
+					 enum at_cmd_state *state)
+{
+	int retval;
+	bool cmee_activated = false;
+
+	int cmee_active_result = cmee_active();
+
+	if (cmee_active_result < 0) {
+		return -EFAULT;
+	}
+
+	if (cmee_active_result == 0) {
+		int result = cmee_set(true);
+
+		if (result != 0) {
+			return result;
+		}
+		cmee_activated = true;
+	}
+
+	retval = at_cmd_write(cmd, buf, buf_len, state);
+
+	if (cmee_activated) {
+		int result = cmee_set(false);
+
+		if (result != 0) {
+			return result;
+		}
+	}
+
+	return retval;
+}
+
+int nrf_inbuilt_key_write(nrf_sec_tag_t sec_tag,
+			  enum nrf_key_mgnt_cred_type_t cred_type,
+			  const u8_t *buffer, u16_t buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u,\"",
+			   NRF_INBUILT_KEY_OP_WR, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	if ((written < 0) || (written > sizeof(scratch_buf))) {
```

Comment:
I'm pretty sure the second condition should be: ` (written >= sizeof(scratch_buf))`.

`written == sizeof(scratch_buf)` condition indicates that the `snprintf` function truncated the null terminator. And we need it as `at_cmd` driver operates on strings.

This would apply in other functions below as well.